upstream: https://codereview.qt-project.org/gitweb?p=pyside%2Fpyside-setup.git;a=commit;h=039776b1af905ab90ab49f6986214450e8099882

details:
https://bugreports.qt.io/browse/PYSIDE-1601

From 039776b1af905ab90ab49f6986214450e8099882 Mon Sep 17 00:00:00 2001
From: Friedemann Kleint <Friedemann.Kleint@qt.io>
Date: Wed, 2 Jun 2021 14:37:22 +0200
Subject: [PATCH] Fix error about not holding the GIL when releasing QImage
 memory

Fix: "Fatal Python error: Python memory allocator called without holding the GIL"
when running sources/pyside6/tests/QtWidgets/qimage_test.py

Acquire the GIL for releasing.

Amends a97698f0a5d011fd206db94cf73dba6353f39b29.

Pick-to: 6.1 5.15
Task-number: PYSIDE-1563
Change-Id: Idf8dbf857f6dde76f80a78f581dacb3cf5fea16b
Reviewed-by: Christian Tismer <tismer@stackless.com>
---
 sources/pyside6/PySide6/glue/qtgui.cpp | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/sources/pyside6/PySide6/glue/qtgui.cpp b/sources/pyside6/PySide6/glue/qtgui.cpp
index 681539c97..ef5caa545 100644
--- a/sources/pyside6/PySide6/glue/qtgui.cpp
+++ b/sources/pyside6/PySide6/glue/qtgui.cpp
@@ -163,7 +163,10 @@ for (Py_ssize_t i = 0; i < count; ++i){
 // @snippet qimage-decref-image-data
 static void imageDecrefDataHandler(void *data)
 {
+    // Avoid "Python memory allocator called without holding the GIL"
+    auto state = PyGILState_Ensure();
     Py_DECREF(reinterpret_cast<PyObject *>(data));
+    PyGILState_Release(state);
 }
 // @snippet qimage-decref-image-data
 
-- 
2.16.3
